oteapi/plugins/factories.py

```diff
@@ -104,70 +111,81 @@ class StrategyFactory:
-            configuration, through the `model` parameter.
+            configuration, through the `config` parameter.
 
         """
-
-        try:
-            if not index and field:
-                index = (field, model.dict()[field])
-            elif not index:
-                raise ValueError("field or index must be specified.")
-            retval = cls.strategy_create_func[index]
-        except KeyError as err:
-            raise NotImplementedError(f"{index=!r} does not exist") from err
-        return retval(model)
+        if isinstance(strategy_type, str):
+            try:
+                strategy_type = StrategyType.init(strategy_type)
+            except ValueError as exc:
+                raise ValueError(
+                    f"Strategy type {strategy_type!r} is not supported."
+                ) from exc
+        elif not isinstance(strategy_type, StrategyType):
+            raise TypeError(
+                "strategy_type should be either of type StrategyType or a string."
+            )
+
+        strategy_name: str = cls._get_strategy_name(config, strategy_type)
+
+        if strategy_name in cls.strategy_create_func[strategy_type]:
+            return cls.strategy_create_func[strategy_type][
+                strategy_name
+            ].implementation(config)
+        raise NotImplementedError(
+            f"The {strategy_type.value} strategy {strategy_name!r} does not exist."
+        )
 
     @classmethod
-    def register(
-        cls, *args: "Tuple[str, ValueType]"
-    ) -> "Callable[[Any], Type[IStrategy]]":
-        """Register a strategy.
+    def _get_strategy_name(
+        cls,
+        config: "StrategyConfig",
+        strategy_type: StrategyType,
+    ) -> str:
+        """Return the strategy name through the config.
 
-        The identifier for the strategy is defined by a set of key-value tuple pairs.
-        """
+        This is a method to accommodate strategy type-specific quirks to retrieve the
+        strategy name.
 
-        def decorator(strategy_class: "Type[IStrategy]") -> "Type[IStrategy]":
-            for index in args:
-                if index not in cls.strategy_create_func:
-                    print(f"Registering {strategy_class.__name__} with {index}")
-                    cls.strategy_create_func[index] = strategy_class
-                else:
-                    raise KeyError(f"{index=!r} is already registered.")
-            return strategy_class
+        Parameters:
+            config: A strategy configuration.
+            strategy_type: The strategy type as initialized in `make_strategy()`.
 
-        return decorator
+        Returns:
+            The strategy name provided in the configuration.
 
-    @classmethod
-    def unregister(cls, *args: "Tuple[str, ValueType]") -> None:
-        """Unregister a strategy."""
-        for index in args:
-            cls.strategy_create_func.pop(index, None)
+        """
+        if strategy_type == StrategyType.DOWNLOAD:
```

Comment:
This looks like a hack to me. Why do you have to do a special trick for the download strategy here?

Reply:
This was also done previously.
Unfortunately, where the strategy type + name is derived directly from a key + value in the configuration normally, this is not the case for the download strategy, where the strategy type (`scheme`) is a special attribute of the `downloadUrl` configuration key. The strategy name is the value of this `scheme` attribute.

Reply:
I have created this special extra method to deal with these edge cases - instead of making it a proper "hack" by simply implementing it directly in the `make_strategy()` method.

Therefore, should any other strategy-specific quirks appear in the future, they can be added to this special (private) method.

Reply:
I think this is a good way to do it.